Add publish request support for multiple responses

ColinSullivan1 opened the pull request:
Adds a publish request wait to allow for multiple responses from a publish call.  Updated to honor the timeout for the first message, then wait a `reply-timeout` for subsequent messages, a suggestion by @derekcollison for better UX.

`--replies 0` will gather responses until the global timeout is reached.

### Example

```text
nats reply foo "response from service a" --queue a &
nats reply foo "response from service b" --queue b &
nats reply foo "response from service c" --queue c &
```

Publish and wait (edited)

```
$ nats req foo hello
11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.Wu6eONq021mLlGG5nWuMGp.SqEPt48b" rtt 406.868µs
response from service a

$ nats req --replies 2 foo hello
11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.qf0sz6O8QzaQR3hp2NEfIC.AQNxkIEZ" rtt 277.961µs
response from service b

11:15:03 Received on "_INBOX.qf0sz6O8QzaQR3hp2NEfIC.AQNxkIEZ" rtt 292.951µs
response from service c

$ nats req --count 2 --replies 2 foo hello
11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.CxwBcyiUTWqGKVDNTRhivV.zgA2BcQD" rtt 286.797µs
response from service b

11:15:03 Received on "_INBOX.CxwBcyiUTWqGKVDNTRhivV.zgA2BcQD" rtt 332.823µs
response from service c

11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.CxwBcyiUTWqGKVDNTRhivV.QCt1hPjN" rtt 204.034µs
response from service c

11:15:03 Received on "_INBOX.CxwBcyiUTWqGKVDNTRhivV.QCt1hPjN" rtt 212.46µs
response from service b

$ nats req --replies 4 foo hello
11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.xbJ7N0Fn7DjrqGhUncDuJi.03IZvLTk" rtt 283.41µs
response from service c

11:15:03 Received on "_INBOX.xbJ7N0Fn7DjrqGhUncDuJi.03IZvLTk" rtt 296.587µs
response from service b

11:15:03 Received on "_INBOX.xbJ7N0Fn7DjrqGhUncDuJi.03IZvLTk" rtt 300.663µs
response from service a

$ nats req --replies 0 --timeout 2s foo hello
11:15:03 Sending request on "foo"
11:15:03 Received on "_INBOX.ed9PxZ7jHPwOcAV6vivU2s.0zCVrCFU" rtt 334.656µs
response from service c

11:15:03 Received on "_INBOX.ed9PxZ7jHPwOcAV6vivU2s.0zCVrCFU" rtt 350.819µs
response from service a

11:15:03 Received on "_INBOX.ed9PxZ7jHPwOcAV6vivU2s.0zCVrCFU" rtt 355.099µs
response from service b

$ nats req --replies 4 --reply-timeout 1s foo hello
11:15:05 Sending request on "foo"
11:15:05 Received on "_INBOX.eUkVIs3ld62Uu1LzLtrrCR.CFE2ojkB" rtt 286µs
response from service c

11:15:05 Received on "_INBOX.eUkVIs3ld62Uu1LzLtrrCR.CFE2ojkB" rtt 303.916µs
response from service b

11:15:05 Received on "_INBOX.eUkVIs3ld62Uu1LzLtrrCR.CFE2ojkB" rtt 310.99µs
response from service a

$ nats req --replies 4 --reply-timeout 1us foo hello
11:15:06 Sending request on "foo"
11:15:06 Received on "_INBOX.7uSdKcRuPFxBsIEq3KTVdb.uCUmA2bZ" rtt 286.172µs
response from service a

11:15:06 Received on "_INBOX.7uSdKcRuPFxBsIEq3KTVdb.uCUmA2bZ" rtt 300.743µs
response from service b

11:15:06 Received on "_INBOX.7uSdKcRuPFxBsIEq3KTVdb.uCUmA2bZ" rtt 304.645µs
response from service c

$ nats req --count 2 --replies 0 --timeout 2s foo hello
11:15:06 Sending request on "foo"
11:15:06 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.vXN8TArl" rtt 298.671µs
response from service a

11:15:06 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.vXN8TArl" rtt 315.217µs
response from service b

11:15:06 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.vXN8TArl" rtt 319.691µs
response from service c

11:15:08 Sending request on "foo"
11:15:08 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.79AY5Nin" rtt 356.559µs
response from service c

11:15:08 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.79AY5Nin" rtt 369.079µs
response from service b

11:15:08 Received on "_INBOX.lASTUVgiH8vp0WjGItD2FZ.79AY5Nin" rtt 373.056µs
response from service a

$ nats req --timeout 2s invalid hello
11:15:10 Sending request on "invalid"
11:15:10 No responders are available

$ nats req --timeout 1us --count 4 foo hello
11:15:10 Sending request on "foo"
11:15:10 Sending request on "foo"
11:15:10 Sending request on "foo"
11:15:10 Sending request on "foo"
```

Resolves #403 

Signed-off-by: Colin Sullivan <colin@synadia.com>

tbeets commented:
Would be cool to add max receives to max duration (whichever happens first).  

Here's an example riffing on your PR: https://github.com/tbeets/gonats-101/blob/main/nats-req-multi/main.go

ColinSullivan1 commented:
Yep, cool - or thinking about it, could expand this with a count and/or handle timeout... e.g. https://gist.github.com/r4um/c1ab51b8757fc2d75d30320933cdbdf6

ripienaar commented on cli/pub_command.go:
```diff
@@ -129,6 +131,73 @@ func (c *pubCmd) prepareMsg(body []byte, seq int) (*nats.Msg, error) {
 	return msg, parseStringsToMsgHeader(c.hdrs, seq, msg)
 }
 
+func (c *pubCmd) doReqWait(nc *nats.Conn, progress *uiprogress.Bar) error {
```
in order to decrease the duplication I think we can always use the same req - this one - we just need to know to exit after first message when wait duration is 0

The existing `doReqAsync()` function has a lot of this logic in it, including for example early short circuit when no responses are recieved for 300ms etc. 

So after @tbeets opened the issue my thought was that I would make that a bit more generic - or extract from it it's core 1:n reply handling to use here.  Thats a bit of a chore, happy to do that on Friday.

ColinSullivan1 replied:
I'll take a look, thanks @ripienaar.

ColinSullivan1 commented:
After some thought I simplified the request a bit, de-duped code based on feedback, and used a select like @tbeets suggested.  Also removed one parameter so we're just left with `--wait-multiple`, and the timeout parameter is used instead of wait-duration.  Zero will wait until the timeout.  No responders are also covered.

I looked into reusing the code in util, that would be major work as you mentioned @ripienaar.  This seemed simpler and easier given support of the pub count and sleep options in the meantime.

ripienaar commented on cli/pub_command.go:
```diff
@@ -66,7 +67,8 @@ Available template functions are:
 	pub.Arg("body", "Message body").Default("!nil!").StringVar(&c.body)
-	pub.Flag("wait", "Wait for a reply from a service").Short('w').BoolVar(&c.req)
+	pub.Flag("wait", "Wait for a single reply from a service").Short('w').BoolVar(&c.req)
+	pub.Flag("wait-multiple", "Wait for multiple replies from services. 0 waits until timeout").Default("1").IntVar(&c.waitCount)
```
Wouldnt it be more logical if `nats req` just had an option to wait for multiple responses or a set period keeping `nats pub` as sending only? Having `nats pub` gain some `req` features that `req` doesnt have seems weird to me. 

While these 2 just reuse the same logic and is so close together the line between publish only and interact with services is pretty strong so would want to keep that

Rest of the PR is great thank you

ripienaar replied:
Yes I have the delta code in tbe other function I mentioned. If we need to go all in and do that might still be worth making that a bit more generic for reuse

tbeets replied:
@ColinSullivan1 The `--wait-multiple` option on `nats pub` was elegant.  What are you thinking for `nats req` -- something like just `--multiple`?

ripienaar replied:
Don’t think delta based early termination makes sense here. It makes sense for server api interactions like varz because we know servers respond immediately. We do not know that for some random service.

ColinSullivan1 replied:
I went ahead and added some very simple delta-like code upon request, and that behavior can be overridden with `--request 0`.  Certainly open for discussion or suggestions.  Also moved to the sync sub - it was simpler.

ColinSullivan1 replied:
I updated the description to reflect the new usage, and the timestamps give an indication of UX.

derekcollison commented:
I will let @ripienaar make the call on the code but looks pretty good to me, well done.

ripienaar commented on cli/pub_command.go:
```diff
@@ -66,7 +68,7 @@ Available template functions are:
 	pub.Arg("subject", "Subject to subscribe to").Required().StringVar(&c.subject)
 	pub.Arg("body", "Message body").Default("!nil!").StringVar(&c.body)
-	pub.Flag("wait", "Wait for a reply from a service").Short('w').BoolVar(&c.req)
+	pub.Flag("wait", "Wait for a single reply from a service").Short('w').BoolVar(&c.req)
```
we should probably just remove this option entirely, think its here as a old leftover or just a mistake. pub should just pub.

ripienaar commented on cli/pub_command.go:
```diff
@@ -115,6 +117,9 @@ Available template functions are:
 	req.Flag("header", "Adds headers to the message").Short('H').StringsVar(&c.hdrs)
 	req.Flag("count", "Publish multiple messages").Default("1").IntVar(&c.cnt)
+	req.Flag("replies", "Wait for multiple replies from services. 0 waits until timeout").Default("1").IntVar(&c.replyCount)
+	req.Flag("reply-timeout", "Maximum timeout between incoming replies.").Default("250ms").DurationVar(&c.replyTimeout)
```
If we are making this default we need to somehow explain to users that we are using this and not --timeout in the UX of the application.

As I mentioned earlier I think for interacting with nats server APIs this would be a good default - they respond immediately - but we have no idea what services users interact with and this default could be confusing them not knowing why the default timeout is being ignored for example. Even the example in the docs again wttr using nats reply would not be reliable since that API is far latency wise and might not respond immediately.  

So not sure, we should maybe warn at the end when early terminating based on prefering that value over --timeout

ColinSullivan1 replied:
Per our slack convo, I'll update the subsequent timeout to do 300ms + response time for the first.  So, for example if the first response came in at 100ms, then we'd wait 400ms (100ms + 300ms) for the 2nd, resulting a total of 500ms for the entire operation.

ColinSullivan1 replied:
Update, will use a rtt average.

ripienaar commented on cli/pub_command.go:
```diff
@@ -130,13 +135,13 @@ func (c *pubCmd) prepareMsg(body []byte, seq int) (*nats.Msg, error) {
 		}
 
-		body, err := pubReplyBodyTemplate(c.body, i)
+		body, err := pubReplyBodyTemplate(c.body, 0)
```
I think the `0` here should be `i`, what was the reason for the change? 

This is the value that `{{Count}}` would set in the body template and its the number of messages sent, setting this to 0 always will break that feature.

ColinSullivan1 replied:
That was a bug from an earlier iteration - thanks for catching that - will fix.

ripienaar commented on cli/pub_command.go:
```diff
@@ -146,16 +151,43 @@ func (c *pubCmd) doReq(nc *nats.Conn, progress *uiprogress.Bar) error {
+				}
+
+				log.Printf("Error receiving message: %v", err)
+				return err
```
this will log here and in the exit handler wont it? Error would be logged twice? maybe better `return fmt.Error("could not receive message: %v", err)` which would then be handled and logged higher up

ripienaar reviewed (changes requested):
some minor tweeks, looking good!

ColinSullivan1 commented:
I've added some dynamic averages around rtt for slower systems, so we can have a good UX for systems that respond quickly, and work with systems that have larger RTTs.  If there is a single service that is very slow to respond then that response may be lost.  However, that is also what `--replies 0` with a long timeout would handle.

ColinSullivan1 commented on cli/pub_command.go:
```diff
@@ -146,39 +151,87 @@ func (c *pubCmd) doReq(nc *nats.Conn, progress *uiprogress.Bar) error {
+			}
+
+			rtt := time.Since(start)
+			log.Printf("Received on %q rtt %v", m.Subject, rtt)
```
It was in the original request API.  I can go either way - wdyt?

ripienaar replied:
ok we leave as is for now :)

ripienaar reviewed (approved):
LGTM
